feat: add spot illustration support into dialtone

braddialpad opened the pull request:
## Description
Spot illustration support. 

Themable spot illustrations can be added by using the classes `d-svg-primary--stroke` or `d-svg-primary--fill` on their path tags within the svg. any definitions of these specific classes within <style> tags in the local file will be removed on build, making it use the primary-color class instead. This way the primary-color of the illustration will change as the theme changes.

## Pull Request Checklist

 - [x] Ask the contributors if a similar effort is already in process or has been solved.
 - [x] Review the [contribution guidelines](https://github.com/dialpad/dialtone/blob/staging/.github/CONTRIBUTING.md).
 - [x] Use `staging` as your pull request's base branch. (All PRs using `production` as its base branch will be declined).
 - [x] Ensure all `gulp` scripts successfully compile.
 - [x] Update, remove, or extend all affected documentation.
 - [x] Ensure no private Dialpad links or info are in the code or pull request description (Dialtone is a public repo!).

## Obligatory GIF (super important!)
![giphy (1)](https://user-images.githubusercontent.com/64808812/141569689-b690c8c3-4607-4aed-a4ec-e1857e917f61.gif)
)

nikitadialpad commented:
hey Brad, it's some code alignment thing [here](https://deploy-preview-502--dialtone.netlify.app/design/illustrations/spot/). Could indentation be aligned like [it's on input page](https://deploy-preview-502--dialtone.netlify.app/components/input/)?

dialpad-drew commented on docs/design/illustrations/spot.html:
```diff
@@ -0,0 +1,79 @@
+</div>
+<section class="d-stack16 d-px16">
+  {% header "h2", "Theming" %}
+  {% paragraph %}Spot illustrations can use the current theme's primary color within the SVG. In order to do this you must set one of the following classes on your SVG path tags:{% endparagraph %}
```
I read this as the assets _aren't_ theme-able by default, but can be _made_ theme-able by adding these classes into the asset. Is that intended?

braddialpad replied:
That's correct. Is it a possibility that some spot illustrations would look the same regardless of theme (say if the illustration was black and white) or do we want to enforce that all spot illustrations are themeable?

dialpad-drew replied:
If the spot illustration style uses our primary color (or derivates of our primary color), I'd suggest we make it theme-able by default. If not, I fear it undermines our claim that the system is theme-able and I'm not sure I see a benefit in making people manually manage SVG XML. Seems like it causes us to write more documentation and our consuming teams to read more documentation.

One could make the argument that Dialtone could provide an SVG component and we just start handling theme-ing in there. 😇

braddialpad replied:
simple enough to do this if that's what we want, we can just replace those colors with the class on build.

dialpad-drew commented on docs/design/illustrations/spot.html:
```diff
@@ -0,0 +1,79 @@
+  {% header "h2", "Theming" %}
+  {% paragraph %}Spot illustrations can use the current theme's primary color within the SVG. In order to do this you must set one of the following classes on your SVG path tags:{% endparagraph %}
+  <ul class="d-stack8">
+    <li><strong>d-svg-primary--stroke</strong> - to set the stroke to the theme's primary color</li>
+    <li><strong>d-svg-primary--fill</strong> - to set the fill to the theme's primary color</li>
```
Might recommend using this shortcode for inline code like the classes on these lines:  `{% code %}foo{% endcode %}`

dialpad-drew commented on docs/design/illustrations/spot.html:
```diff
@@ -0,0 +1,79 @@
+  {% paragraph %}We also recommend using opacity to create other colors derived from the theme's primary color where possible.{% endparagraph %}
+  {% codeWell %}
+    {% codeWellHeader %}
+      In the below example we are using d-svg-primary--fill and d-svg-primary--stroke as a css class within our SVG. If you were to view the SVG locally, it would use the set color #3dffab. However once this is built into dialtone the local style tags would be removed and it would use the primary color of the dialtone theme instead.
```
`codeWellHeader` is typically for visualizing an example related to the code snippet. Is there a reason we're putting this copy in there and not with the other paragraphs? 

Also, see above about the `{% code%}` shortcode -- would recommend using it with the classes here.

braddialpad replied:
I just wanted it to be associated with the example, kind of like a caption, but I should probably change that since it's non standard across the rest of the site

dialpad-drew commented on docs/design/illustrations/spot.html:
```diff
@@ -0,0 +1,79 @@
+    {% endcodeWellHeader %}
+    {% codeWellFooter %}
+      {% highlight html linenos %}
+      <svg width="96" height="96" viewBox="0 0 96 96" fill="none" xmlns="http://www.w3.org/2000/svg">
+        <style>
+          .d-svg-primary--stroke {
+              stroke: #3dffab;
+          }
+
+          .d-svg-primary--fill {
+              fill: #3dffab;
+          }
+        </style>
+        <path class="d-svg-primary--fill"
+          d="M48 96C74.5097 96 96 74.5097 96 48C96 21.4903 74.5097 0 48 0C21.4903 0 0 21.4903 0 48C0 74.5097 21.4903 96 48 96Z" />
+        <path class="d-svg-primary--stroke"
+          d="M48 96C74.5097 96 96 74.5097 96 48C96 21.4903 74.5097 0 48 0C21.4903 0 0 21.4903 0 48C0 74.5097 21.4903 96 48 96Z" />
+      </svg>
```
What is this code snippet an example of? Theming? How a spot illustration would be implemented into UC/UV?

Also, nit: for code snippets, I'd recommend shift-tabbing so the left-most element is flush with the text editor window. These snippets area wrapped with `<pre>` and can preserve some funky spacing.

braddialpad replied:
yeah the indenting is messed, thanks for the eye.

this is an example of how to make your svg themeable when adding it to dialtone. I'm thinking maybe this should belong to the "adding icons" section of the site rather than here.

dialpad-drew reviewed (commented):
How does one _add_ a spot illustration to our project? I assume the SVG should be added to the project and the JSON file. Are there any additional or new gulp tasks that need to be run?

braddialpad commented:
> How does one _add_ a spot illustration to our project? I assume the SVG should be added to the project and the JSON file. Are there any additional or new gulp tasks that need to be run?

I see there is an adding icons page that splits icons into system/brand/patterns to some extent already. I'll add spot illustrations to here as well.

The process for adding a spot illustration is identical other than setting the `d-svg-primary--fill` and `d-svg-primary--stroke` classes if you want to use theme primary colors

braddialpad commented:
I've updated the gulp process to use the dialpad primary color for theming rather than a css class. 

Have also updated the adding icons page to include spot illustrations and added a section on theming. 

The spot illustrations page now links to the adding icons page to explain how to theme spot illustrations.

I believe all outstanding comments should be resolved.

dialpad-drew commented on docs/design/illustrations/spot.html:
```diff
@@ -0,0 +1,50 @@
+              </div>
+              <p class="dialtone-icon-card__subtitle">{{ name }}</p>
+            </header>
+            <footer class="dialtone-icon-card__footer js-dialtone-icon-card-footer">
```
Is it possible for just the spot illustration page's `.dialtone-icon-card__footer` to have `width: calc(var(--icon-card-width)*1.5)` set on it? Its current width is based on the asset being smaller icons, so with these assets being larger, it makes this pop-up area abnormally large. I'll show you a before and after:

## Before
<img width="633" alt="Screen Shot 2021-11-17 at 4 16 03 PM" src="https://user-images.githubusercontent.com/77751297/142297936-51e1e9d2-ec5d-41dc-b5ac-58597f31986d.png">

## After
<img width="548" alt="Screen Shot 2021-11-17 at 4 16 06 PM" src="https://user-images.githubusercontent.com/77751297/142297954-539efe00-1fe5-4030-b2e9-b9dfc61b1308.png">

dialpad-drew commented on docs/design/icons/adding.html:
```diff
@@ -10,6 +10,11 @@
     <li><strong>Can I use another icon already in Dialtone to achieve the same purpose?</strong><br/>Every icon introduced is yet another thing that our users must parse and learn. Whenever possible, consider using a pre-existing icon that captures the essence of what you are trying to communicate and clarify further with a label.</li>
   {% endol %}
 </section>
+<section class="d-stack16">
+  {% paragraph %}<strong>A note on theming</strong>{% endparagraph %}
+  {% paragraph %}Spot Illustrations allow for theming, meaning the primary color of the illustration will change when the theme changes. If your spot illustration svg contains a fill or a stroke with the dialtone primary color {% code %}#6C3DFF{% endcode %} then it will automatically change when the theme's primary color changes.{% endparagraph %}
+  {% paragraph %}You may often wish to derive lighter or darker colors based on the primary color when making a spot illustration. In these instances it is recommended to set opacity while still using the primary color as the base color so these colors will also change with the theme.{% endparagraph %}
+</section>
```
I might recommend adding this _under_ the instructions. My theory is the instructions is the primary content people will seek more often and the theming bit is more secondary.

dialpad-drew reviewed (approved):
Two small things. Nice job, @braddialpad !

braddialpad commented:
> I might recommend adding this _under_ the instructions. My theory is the instructions is the primary content people will seek more often and the theming bit is more secondary. 

yep I agree, moved this to the bottom

braddialpad commented:
> Is it possible for just the spot illustration page's `.dialtone-icon-card__footer` to have `width: calc(var(--icon-card-width)*1.5)` set on it? Its current width is based on the asset being smaller icons, so with these assets being larger, it makes this pop-up area abnormally large. I'll show you a before and after:

done!

juliodialpad commented on packages/dialtone-tokens/postcss/dialtone-tokens.cjs:
```diff
@@ -73,17 +73,18 @@ function boxShadows (shadowDeclarations, Declaration) {
         }).join(', ');
 
-      shadowDeclarations.at(-1).after(new Declaration({ prop: shadowVar, value }));
+      shadowDeclarations.at(0).after(new Declaration({ prop: shadowVar, value }));
```
Why this was -1 previously? 🤔

braddialpad replied:
I had to reverse the array to get it to work correctly in the shadow parsing function after recent changes, and I thought if I put it at the end of the array it would output at the bottom, but actual array order doesn't seem to affect the order postcss uses to output (probably uses a linked list or something internally). This will now output after the shadow declarations instead of the top of the list, so it makes more logical sense when you are looking at the file.

No functional difference since order of css vars doesn't matter logic wise.

juliodialpad reviewed (approved):
Fix looks good, just one question, thanks!
